File: backend/internal/handler/analyze.go
package handler

import (
	"context"
	"net/http"

	"github.com/labstack/echo/v5"
)

// AnalyzeUseCase はHandlerが必要とする機能だけを定義する。
//
// Handlerは具体的なAnalyzePullRequestsUseCaseではなく、
// このinterfaceに依存する。
//
// これによりHandlerのテスト時に
// Fake UseCaseへ簡単に差し替えられる。
type AnalyzeUseCase interface {
	Execute(
		ctx context.Context,
		owner string,
		repo string,
	) (string, error)
}

// AnalyzeHandler はPR分析APIのHTTP層を担当する。
//
// GitHubやGeminiなどの具体的な処理は知らず、
// AnalyzeUseCaseを呼び出すことだけを担当する。
type AnalyzeHandler struct {
	usecase AnalyzeUseCase
}

// NewAnalyzeHandler はHandlerを生成する。
func NewAnalyzeHandler(
	uc AnalyzeUseCase,
) *AnalyzeHandler {
	return &AnalyzeHandler{
		usecase: uc,
	}
}

// AnalyzeRequest はPOST /api/analyzeの入力DTO。
type AnalyzeRequest struct {
	Owner string `json:"owner"`
	Repo  string `json:"repo"`
}

// AnalyzeResponse は正常時のレスポンスDTO。
type AnalyzeResponse struct {
	Result string `json:"result"`
}

// ErrorResponse はエラー時のレスポンスDTO。
type ErrorResponse struct {
	Message string `json:"message"`
}

// Analyze はPR分析APIを処理する。
func (h *AnalyzeHandler) Analyze(c *echo.Context) error {
	var request AnalyzeRequest

	// JSON BodyをリクエストDTOへ変換する。
	if err := c.Bind(&request); err != nil {
		return c.JSON(
			http.StatusBadRequest,
			ErrorResponse{
				Message: "リクエスト形式が不正です。",
			},
		)
	}

	// MVPではownerとrepoを必須とする。
	if request.Owner == "" || request.Repo == "" {
		return c.JSON(
			http.StatusBadRequest,
			ErrorResponse{
				Message: "owner と repo は必須です。",
			},
		)
	}

	// HTTP RequestのContextをそのままUseCaseへ渡す。
	result, err := h.usecase.Execute(
		c.Request().Context(),
		request.Owner,
		request.Repo,
	)
	if err != nil {
		// 内部エラーの詳細はクライアントへ露出させない。
		return c.JSON(
			http.StatusInternalServerError,
			ErrorResponse{
				Message: "Pull Requestの分析に失敗しました。",
			},
		)
	}

	return c.JSON(
		http.StatusOK,
		AnalyzeResponse{
			Result: result,
		},
	)
}

File: frontend/src/lib/api/analyze.ts
// Go Backendへ送信するRequest型。
export type AnalyzeRequest = {
  owner: string;
  repo: string;
};

// Go Backendから返ってくるResponse型。
export type AnalyzeResponse = {
  result: string;
};

// Go Backendでエラーになった場合のResponse型。
type ErrorResponse = {
  message: string;
};

// GitHub Repositoryを分析するAPI。
//
// UI Componentから直接fetchを書かず、
// API通信処理をこのファイルへまとめる。
//
// 将来的にAPI URLや認証方式が変わっても、
// UI側への影響を小さくする。
export async function analyzeRepository(
  request: AnalyzeRequest,
): Promise<AnalyzeResponse> {
  // 環境変数からGo BackendのURLを取得する。
  //
  // NEXT_PUBLIC_ を付けることで、
  // Browser側のJavaScriptから参照できる。
  const apiUrl =
    process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:8080";

  // Go Backendへ分析リクエストを送信する。
  const response = await fetch(`${apiUrl}/api/analyze`, {
    method: "POST",

    headers: {
      "Content-Type": "application/json",
    },

    body: JSON.stringify(request),
  });

  // HTTP 4xx / 5xxの場合。
  if (!response.ok) {
    // Backendが返したエラーメッセージを取得する。
    //
    // JSONとして取得できなかった場合も考慮して
    // catchでデフォルトメッセージへフォールバックする。
    const errorResponse: ErrorResponse = await response
      .json()
      .catch(() => ({
        message: "分析に失敗しました。",
      }));

    throw new Error(errorResponse.message);
  }

  // 正常時のJSONレスポンスを返す。
  return response.json() as Promise<AnalyzeResponse>;
}